lib/nvme/nvme_transport.c

```diff
@@ -513,26 +603,41 @@ nvme_transport_ctrlr_connect_qpair(struct spdk_nvme_ctrlr *ctrlr, struct spdk_nv
 		goto err;
 	}
 
-	if (qpair->poll_group) {
+	/* Skip TCP I/O queue pair registration since it's already handled in
+	 * nvme_transport_connect_qpair.
+	 */
+	if (qpair->poll_group && (ctrlr->trid.trtype != SPDK_NVME_TRANSPORT_TCP)) {
 		rc = nvme_poll_group_connect_qpair(qpair);
 		if (rc) {
 			goto err;
 		}
 	}
 
 	if (!qpair->async) {
-		/* Busy wait until the qpair exits the connecting state */
-		while (nvme_qpair_get_state(qpair) == NVME_QPAIR_CONNECTING) {
-			if (qpair->poll_group && spdk_nvme_ctrlr_is_fabrics(ctrlr)) {
-				rc = spdk_nvme_poll_group_process_completions(
-					     qpair->poll_group->group, 0,
-					     nvme_transport_connect_qpair_fail);
-			} else {
-				rc = spdk_nvme_qpair_process_completions(qpair, 0);
+		if (ctrlr->trid.trtype != SPDK_NVME_TRANSPORT_TCP) {
+			/* Busy wait until the qpair exits the connecting state */
+			while (nvme_qpair_get_state(qpair) == NVME_QPAIR_CONNECTING) {
+				if (qpair->poll_group && spdk_nvme_ctrlr_is_fabrics(ctrlr)) {
+					rc = spdk_nvme_poll_group_process_completions(
+							qpair->poll_group->group, 0,
+						    nvme_transport_connect_qpair_fail);
+				} else {
+					rc = spdk_nvme_qpair_process_completions(qpair, 0);
+				}
+
+				if (rc < 0) {
+					goto err;
+				}
 			}
-
-			if (rc < 0) {
-				goto err;
+		} else {
+			/*
+			 * Non-blocking path for TCP transport. Use a poller to complete connection
+			 * asynchronously.
+			 */
+			rc = start_async_qpair_connect(qpair, nvme_qpair_connect_completion_cb, NULL);
+			if (rc != 0) {
+			    SPDK_ERRLOG("Failed to start async connect: %d\n", rc);
+			    goto err;
```

Comment:
Sorry if my question does not make sense. Does it mean we also use poller for TCP transport even if in non-interrupt mode?

Reply:
Got it. Thanks @c3y1huang !